### ISSUE TITLE
Restart IoT or edge tasks before reconnecting to server - feedback

### ISSUE THREAD

**Issue by nickwales** [hashicorp/nomad#15783]
With the launch of [Nomad 1.3](https://www.hashicorp.com/blog/nomad-1-3-adds-native-service-discovery-and-edge-workload-support), we added the ability to determine how long edge Nomad clients could be unavailable for before rescheduling their tasks.

This adds a lot of benefit for many edge workloads, however we have identified that for some IoT applications the power and connectivity of the Nomad client is out of the control of the service provider. In these cases an outage on the edge client would require connectivity back to the Nomad servers to resume before a task could be rescheduled, potentially increasing the time to recovery. 

To overcome this it would be beneficial for the Nomad client to have some additional autonomous functionality that would attempt to restart the previously running tasks immediately upon resumption. 

We would love to hear if this feature set would improve the stability of your IoT or other edge workloads!

What else would you want from Nomad's edge capabilities? 

Please let us know!

**Comment by lgfa29** [comment]
Thanks @nickwales!

I discussed this with the rest of the team and we came to the conclusion that this should be considered a bug in the current implement that causes edge clients to have an undesired behaviour.

To clarify the issue described here, clients running allocations with [`max_client_disconnect`](https://developer.hashicorp.com/nomad/docs/job-specification/group#max_client_disconnect) set have their workloads migrated to another client when they miss heartbeats.

The allocations keep running while the client is disconnected, but if the client machine reboots the Nomad agent will [wait indefinitely until it is able to contact the servers](https://github.com/hashicorp/nomad/blob/v1.4.3/client/allocrunner/taskrunner/task_runner.go#L551) to recover its workloads.

This behaviour was originally introduce to solve https://github.com/hashicorp/nomad/issues/1795, where old allocation were being incorrectly restarted. But edge workloads should restart even when the server cannot be reached, since this is behaviour users expect when using `max_client_disconnect`.
